### PR TITLE
Instrument command service queues [KVL-519]

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by SamirTalwar-DA — APPROVED** [review]
I thought I posed these comments earlier but forgot to hit the button. Sorry. :slightly_frowning_face:

File: libs-scala/akka-stream-util/src/main/scala/com/daml/akka/stream/InstrumentedSource.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    * For this reason, you may also read values on the saturation counter which
+    * are negative or exceed `bufferSize`.
+    *
+    * Note that the fact that the count is decremented in a second operator means
+    * that its buffering will likely skew the measurements to be greater than the
+    * actual value, rather than the other way around.
+    *
+    * FIXME The clean (but involved) thing to do, would be to re-implement
+    * FIXME `Source.queue` from scratch, adding the possibility of
+    * FIXME instrumenting it.
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply libs-scala/akka-stream-util/src/main/scala/com/daml/akka/stream/InstrumentedSource.scala:58]: Yep I agree this is "good enough" :+1:

File: libs-scala/akka-stream-util/src/test/scala/com/daml/akka/stream/InstrumentedSourceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        case _ => false
+      }
+      inputSize shouldEqual (enqueued + dropped)
+      assert(enqueued >= bufferSize)
+      assert(dropped <= bufferSize)
+      assert(maxBuffered.getCount >= lowAcceptanceThreshold)
+      assert(maxBuffered.getCount <= highAcceptanceThreshold)
+    }
+  }
+
```
Comment (dajmaki) [inline libs-scala/akka-stream-util/src/test/scala/com/daml/akka/stream/InstrumentedSourceSpec.scala:94]: Would be good to also have a test that checks that `Counter.dec()` gets called.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 3]
```diff
@@ -36,6 +36,15 @@ final class Metrics(val registry: MetricRegistry) {
         registry.meter(Prefix :+ "delayed_submissions")
       val validSubmissions: Meter =
         registry.meter(Prefix :+ "valid_submissions")
+
+      def inputBufferSaturation(party: String): Counter =
```
Comment (dajmaki) [inline ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:40]: Does it make sense to have these per-party?

File: libs-scala/akka-stream-util/src/main/scala/com/daml/akka/stream/InstrumentedSource.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import com.codahale.metrics.Counter
+import com.daml.dec.DirectExecutionContext
+
+import scala.concurrent.Future
+
+object InstrumentedSource {
+
+  final class QueueWithComplete[T](
+      delegate: SourceQueueWithComplete[T],
+      saturation: Counter,
```
Comment (dajmaki) [inline libs-scala/akka-stream-util/src/main/scala/com/daml/akka/stream/InstrumentedSource.scala:18]: This isn't saturation in the sense I understand it (% out of total capacity), but rather the number of elements queued.

Comment (dajmaki) [reply libs-scala/akka-stream-util/src/main/scala/com/daml/akka/stream/InstrumentedSource.scala:18]: I've renamed the counters to be "capacity" (was size) and "length" (was saturation).

File: docs/source/tools/sandbox.rst [hunk 5]
```diff
@@ -377,6 +377,30 @@ further sent to deduplication and interpretation.
 A timer. Time to validate submitted commands before they are
 fed to the DAML interpreter.
 
+``daml.commands.<party_name>.input_buffer_size``
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply docs/source/tools/sandbox.rst:380]: Added.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala (hunk 3 above)
Comment (dajmaki) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:40]: I think for now this is reasonable, though I'd appreciate more opinions on this. Do we run into problems with 10k unique parties?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerImpl.scala [hunk 6]
```diff
@@ -76,9 +78,12 @@ private[services] object TrackerImpl {
         Ctx[Promise[Completion], Completion],
         Materialized[NotUsed, Promise[Completion]]],
       inputBufferSize: Int,
+      sizeCounter: Counter,
+      saturationCounter: Counter,
   )(implicit materializer: Materializer, loggingContext: LoggingContext): TrackerImpl = {
-    val ((queue, mat), foreachMat) = Source
-      .queue[QueueInput](inputBufferSize, OverflowStrategy.dropNew)
+    sizeCounter.inc(inputBufferSize.toLong)
```
Comment (dajmaki) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerImpl.scala:84]: This is never decremented and we keep incrementing it every time the tracker gets created. I've refactored the code to handle the inc/dec in InstrumentedSource instead.

File: libs-scala/akka-stream-util/src/test/scala/com/daml/akka/stream/InstrumentedSourceSpec.scala (hunk 2 above)
Comment (dajmaki) [reply libs-scala/akka-stream-util/src/test/scala/com/daml/akka/stream/InstrumentedSourceSpec.scala]: Added test.

File: release/artifacts.yaml [hunk 7]
```diff
@@ -131,6 +131,8 @@
   type: jar-scala
 - target: //libs-scala/auth-utils:auth-utils
   type: jar-scala
+- target: //libs-scala/akka-stream-util:akka-stream-util
+  type: jar-scala
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply release/artifacts.yaml]: Moved the code into ledger/metrics instead as we had similar utilities there.

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/metrics/src/test/scala/com/daml/metrics/InstrumentedSourceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      capacityCounter.getCount shouldEqual 0
+      maxBuffered.decrements.get shouldEqual bufferSize
+    }
+  }
+
+  it should "track the buffer saturation correctly when dropping items" in {
+
+    val bufferSize = 500
+
+    // Due to differences in scheduling, we accept that the highest
```
Comment (hanshoglund-da) [inline ledger/metrics/src/test/scala/com/daml/metrics/InstrumentedSourceSpec.scala:54]:
```suggestion
    // Due to differences in scheduling, we expect the highest
```

**Review by hanshoglund-da — APPROVED** [review]
LGTM

**Pull request merged**